### ISSUE TITLE
server: create visibility into the establishment/teardown of node-to-node connections

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#67662]
( This will be a prerequisite to addressing #54007 )

**tldr/goal**: we want more visibility into the network connection events so that we better understand the impact of future changes in the connection handshake.

### Some expected results

1) more visibility into node-node connections/traffic between node startup and join events. for example:
 
![image](https://user-images.githubusercontent.com/642886/125812679-c44226f2-68b5-4f00-bf8a-c5ed90370070.png)

Here we want to see log messages reporting that the node-node connection has been established, with some properties about the connection (address on the other side, "cluster name" if provided and other fields in the RPC handshake as available)

2) clearer message when "connecting to gossip" does not _actually_ connect to anything, for example in this case:

```
I210715 15:14:22.839832 419 1@server/server.go:1597 ⋮ [n?] 23  connecting to gossip network to verify cluster ID ‹"ef1e1a93-0616-433f-975f-e3f62e5c3825"›
I210715 15:14:22.840169 91 gossip/gossip.go:402 ⋮ [n1] 24  NodeDescriptor set to ‹node_id:1 address:<network_field:"tcp" address_field:"kenax:26257" > attrs:<> locality:<> ServerVersion:<major_val:21 minor_val:1 patch:0 internal:116 > build_tag:
"v21.2.0-alpha.00000000-2403-ga67ea9565b" started_at:1626362062840161455 cluster_name:"" sql_address:<network_field:"tcp" address_field:"kenax:26257" >›
I210715 15:14:22.840346 396 kv/kvserver/closedts/provider/provider.go:135 ⋮ [ct-closer] 25  disabling legacy closed-timestamp mechanism; the new one is enabled
I210715 15:14:22.842765 91 server/node.go:392 ⋮ [n1] 26  initialized store s1
I210715 15:14:22.842812 91 kv/kvserver/stores.go:250 ⋮ [n1] 27  read 0 node addresses from persistent storage
I210715 15:14:22.842829 419 1@server/server.go:1600 ⋮ [n1] 28  node connected via gossip
```

Here we want a log message that indicates that no server were actually connected to.

3) clearer message about node-node connections when a new node is added for example here:

![image](https://user-images.githubusercontent.com/642886/125813091-d6f1650b-944a-40ab-a655-d76c540b6496.png)

We want to see:
- when the network connection is established
- some information about the connection (remote address, cluster name etc as available)

4) more visibility into *when a connection is dropped*  and then also *when a connection is re-established* (also called "re-dial" in the code)

![image](https://user-images.githubusercontent.com/642886/125816069-6ec2792d-87ae-424d-a2dd-6a0d8fc07801.png)

Here we want to see something about the existing connection being torn down, and then something else about a new connection attempt (before the error message).

5) when a connection error occurs, we want to repeat the details about the connections (addresses on both side, but also node IDs, cluster name if available, etc)

Currently they are missing (see last screenshot above)

### General considerations

- all the cases above are related to our `pkg/rpc` package. For context, all the "interesting" node-node traffic is occurring via a gRPC connection. The code to manage those connections is in the `rpc` package. It is likely that all the logging desired above can be added solely within that package.

  hint: understand that the `grpcDialRaw` function are responsible for establishing outwards connections (this is simple)

  hint: where in the code do we see _incoming_ connections pop up? using a gRPC concept called "interceptor" (akin to callback functions). We have a few already installed in the `server/server.go` file, `(*Server).PreStart()`

  so for incoming connections you will likely need to define a new interceptor to log the connection details. (It's possible that gRPC already has logging facilities? We don't know. Will need to investigate!)

- generally the logging will want to report at a minimum the network addresses on both sides of the connections (not just remote) including TCP port number.

- we want logging connection details to report the node ID on both sides *if available* (it may not be available if the cluster is not initialized yet)

- the "cluster name" is a field in the connection code that's configured by the flag `--cluster-name`. By default it is empty. Log only if available.

**Comment by knz** [comment]
Some skills to gain on the way:

- manually start a multi-node cluster
- set up a secure cluster with certs
- understand configuration for logging (ask Jay)
- probably, do the work outlined above using "structured events" so learn about those

**Comment by knz** [comment]
we've added this logging since.
